GoRogue/MapGeneration/Region.cs

```diff
@@ -0,0 +1,374 @@
+        /// <param name="here">The Point to evaluate</param>
+        /// <returns></returns>
+        public bool Contains(Point here)
+        {
+            return InnerPoints.Contains(here);
+        }
+        /// <summary>
+        /// Shifts the entire region by performing Point addition
+        /// </summary>
+        /// <param name="distance">the distance and direction to shift the region</param>
+        /// <returns></returns>
+        public Region Shift(Point distance)
+        {
+            Region region = new Region(Name, SouthEastCorner + distance, NorthEastCorner + distance, NorthWestCorner + distance, SouthWestCorner + distance);
+            foreach (var subRegion in SubRegions)
+            {
+                ((List<Region>) region.SubRegions).Add(subRegion);
+            }
+            return region;
+        }
+
+
+        /// <summary>
+        /// Cuts Points out of a parent region if that point exists in a subregion,
+        /// and cuts Points out of subregions if that point is already in a subregion
+        /// </summary>
+        /// <remarks>
+        /// This favors regions that were added later, like a stack.
+        /// </remarks>
+        public void DistinguishSubRegions()
+        {
+            List<Region> regionsDistinguished = new List<Region>();
+            IEnumerable<Region> regionsReversed = SubRegions;
+            regionsReversed.Reverse();
+            foreach (Region region in regionsReversed)
+            {
+                foreach (Point point in region.Points.Distinct())
+                {
+                    foreach (Region distinguishedRegion in regionsDistinguished)
+                    {
+                        if (distinguishedRegion.Contains(point))
+                        {
+                            while (region.OuterPoints.Contains(point))
+                                ((List<Point>) region.OuterPoints).Remove(point);
+
+                            while (region.InnerPoints.Contains(point))
+                                ((List<Point>) region.InnerPoints).Remove(point);
+                        }
+                    }
+                }
+
+                regionsDistinguished.Add(region);
+            }
+        }
+
+        /// <summary>
+        /// Removes a common point from the OuterPoints of both regions
+        /// and adds it to the Connections of both Regions
+        /// </summary>
+        /// <param name="a">the first region to evaluate</param>
+        /// <param name="b">the second region to evaluate</param>
+        public static void AddConnectionBetween(Region a, Region b)
+        {
+            List<Point> possible = new List<Point>();
+            List<Point> Points = a.OuterPoints.Where(here => b.OuterPoints.Contains(here) && !a.IsCorner(here) && !b.IsCorner(here)).ToList();
+            foreach (Point Point in Points)
+            {
+                possible.Add(Point);
+            }
+            if (possible.Count() <= 2) return;
```

Comment:
Exception, perhaps?  The fact that this has just failed to add a connection point is rather difficult to detect as-is.